// @clerk/ui/dist/no-rhc/components/OrganizationProfile/MembersSearch.js
import { localizationKeys } from "../../localization/localizationKeys.js";
import magnifying_glass_default from "../../icons/magnifying-glass.js";
import { useLocalizations } from "../../localization/makeLocalizable.js";
import { mqu } from "../../styledSystem/breakpoints.js";
import { Spinner } from "../../primitives/Spinner.js";
import { descriptors } from "../../customizables/elementDescriptors.js";
import { Flex, Icon } from "../../customizables/index.js";
import { Animated } from "../../elements/Animated.js";
import { InputWithIcon } from "../../elements/InputWithIcon.js";
import { ACTIVE_MEMBERS_PAGE_SIZE } from "./OrganizationMembers.js";
import { useEffect, useRef } from "react";
import { jsx } from "@emotion/react/jsx-runtime";

//#region src/components/OrganizationProfile/MembersSearch.tsx
const membersSearchDebounceMs = 500;
const MembersSearch = ({ query, value, memberships, onSearchChange, onQueryTrigger }) => {
	const { t } = useLocalizations();
	const debounceTimer = useRef(null);
	const handleChange = (event) => {
		const eventValue = event.target.value;
		onSearchChange(eventValue);
		if (eventValue === "") onQueryTrigger(eventValue);
	};
	function handleKeyUp() {
		if (debounceTimer.current) clearTimeout(debounceTimer.current);
		debounceTimer.current = setTimeout(() => {
			onQueryTrigger(value.trim());
		}, membersSearchDebounceMs);
	}
	useEffect(() => {
		if (!query || !memberships?.data) return;
		if ((memberships?.count ?? 0) <= ACTIVE_MEMBERS_PAGE_SIZE) memberships?.fetchPage?.(1);
	}, [query, memberships]);
	const isFetchingNewData = value && !!memberships?.isLoading && !!memberships.data?.length;
	return /* @__PURE__ */ jsx(Animated, {
		asChild: true,
		children: /* @__PURE__ */ jsx(Flex, {
			sx: {
				width: "50%",
				[mqu.sm]: { width: "auto" }
			},
			children: /* @__PURE__ */ jsx(InputWithIcon, {
				value,
				type: "search",
				autoCapitalize: "none",
				spellCheck: false,
				"aria-label": "Search",
				placeholder: t(localizationKeys("organizationProfile.membersPage.action__search")),
				leftIcon: isFetchingNewData ? /* @__PURE__ */ jsx(Spinner, { size: "xs" }) : /* @__PURE__ */ jsx(Icon, {
					icon: magnifying_glass_default,
					elementDescriptor: descriptors.organizationProfileMembersSearchInputIcon,
					sx: (t$1) => ({ color: t$1.colors.$colorMutedForeground })
				}),
				onKeyUp: handleKeyUp,
				onChange: handleChange,
				elementDescriptor: descriptors.organizationProfileMembersSearchInput
			})
		})
	});
};

//#endregion
export { MembersSearch };
