// @clerk/ui/dist/no-rhc/components/OrganizationProfile/OrganizationProfileRoutes.js
import { useEnvironment } from "../../contexts/EnvironmentContext.js";
import { Route } from "../../router/Route.js";
import { Switch } from "../../router/Switch.js";
import { Protect } from "../../common/Gate.js";
import { useOrganizationProfileContext } from "../../contexts/components/OrganizationProfile.js";
import { CustomPageContentContainer } from "../../common/CustomPageContentContainer.js";
import { OrganizationGeneralPage } from "./OrganizationGeneralPage.js";
import { OrganizationMembers } from "./OrganizationMembers.js";
import { Suspense, lazy } from "react";
import { jsx, jsxs } from "@emotion/react/jsx-runtime";

//#region src/components/OrganizationProfile/OrganizationProfileRoutes.tsx
const OrganizationBillingPage = lazy(() => import(
	/* webpackChunkName: "op-billing-page"*/
	"./OrganizationBillingPage.js"
).then((module) => ({ default: module.OrganizationBillingPage })));
const OrganizationAPIKeysPage = lazy(() => import(
	/* webpackChunkName: "op-api-keys-page"*/
	"./OrganizationAPIKeysPage.js"
).then((module) => ({ default: module.OrganizationAPIKeysPage })));
const OrganizationPlansPage = lazy(() => import(
	/* webpackChunkName: "op-plans-page"*/
	"./OrganizationPlansPage.js"
).then((module) => ({ default: module.OrganizationPlansPage })));
const OrganizationStatementPage = lazy(() => import(
	/* webpackChunkName: "statement-page"*/
	"./OrganizationStatementPage.js"
).then((module) => ({ default: module.OrganizationStatementPage })));
const OrganizationPaymentAttemptPage = lazy(() => import(
	/* webpackChunkName: "payment-attempt-page"*/
	"./OrganizationPaymentAttemptPage.js"
).then((module) => ({ default: module.OrganizationPaymentAttemptPage })));
const OrganizationProfileRoutes = () => {
	const { pages, isMembersPageRoot, isGeneralPageRoot, isBillingPageRoot, isAPIKeysPageRoot, shouldShowBilling, apiKeysProps } = useOrganizationProfileContext();
	const { apiKeysSettings, commerceSettings } = useEnvironment();
	const customPageRoutesWithContents = pages.contents?.map((customPage, index) => {
		const shouldFirstCustomItemBeOnRoot = !isGeneralPageRoot && !isMembersPageRoot && index === 0;
		return /* @__PURE__ */ jsx(Route, {
			index: shouldFirstCustomItemBeOnRoot,
			path: shouldFirstCustomItemBeOnRoot ? void 0 : customPage.url,
			children: /* @__PURE__ */ jsx(CustomPageContentContainer, {
				mount: customPage.mount,
				unmount: customPage.unmount
			})
		}, `custom-page-${customPage.url}`);
	});
	return /* @__PURE__ */ jsxs(Switch, { children: [customPageRoutesWithContents, /* @__PURE__ */ jsxs(Route, { children: [
		/* @__PURE__ */ jsx(Route, {
			path: isGeneralPageRoot ? void 0 : "organization-general",
			children: /* @__PURE__ */ jsx(Switch, { children: /* @__PURE__ */ jsx(Route, {
				index: true,
				children: /* @__PURE__ */ jsx(OrganizationGeneralPage, {})
			}) })
		}),
		/* @__PURE__ */ jsx(Route, {
			path: isMembersPageRoot ? void 0 : "organization-members",
			children: /* @__PURE__ */ jsx(Switch, { children: /* @__PURE__ */ jsx(Route, {
				index: true,
				children: /* @__PURE__ */ jsx(Protect, {
					condition: (has) => has({ permission: "org:sys_memberships:read" }) || has({ permission: "org:sys_memberships:manage" }),
					redirectTo: isGeneralPageRoot ? "../" : "./organization-general",
					children: /* @__PURE__ */ jsx(OrganizationMembers, {})
				})
			}) })
		}),
		commerceSettings.billing.organization.enabled && shouldShowBilling ? /* @__PURE__ */ jsx(Protect, {
			condition: (has) => has({ permission: "org:sys_billing:read" }) || has({ permission: "org:sys_billing:manage" }),
			children: /* @__PURE__ */ jsx(Route, {
				path: isBillingPageRoot ? void 0 : "organization-billing",
				children: /* @__PURE__ */ jsxs(Switch, { children: [
					/* @__PURE__ */ jsx(Route, {
						index: true,
						children: /* @__PURE__ */ jsx(Suspense, {
							fallback: "",
							children: /* @__PURE__ */ jsx(OrganizationBillingPage, {})
						})
					}),
					commerceSettings.billing.organization.hasPaidPlans ? /* @__PURE__ */ jsx(Route, {
						path: "plans",
						children: /* @__PURE__ */ jsx(Suspense, {
							fallback: "",
							children: /* @__PURE__ */ jsx(OrganizationPlansPage, {})
						})
					}) : null,
					/* @__PURE__ */ jsx(Route, {
						path: "statement/:statementId",
						children: /* @__PURE__ */ jsx(Suspense, {
							fallback: "",
							children: /* @__PURE__ */ jsx(OrganizationStatementPage, {})
						})
					}),
					/* @__PURE__ */ jsx(Route, {
						path: "payment-attempt/:paymentAttemptId",
						children: /* @__PURE__ */ jsx(Suspense, {
							fallback: "",
							children: /* @__PURE__ */ jsx(OrganizationPaymentAttemptPage, {})
						})
					})
				] })
			})
		}) : null,
		apiKeysSettings.orgs_api_keys_enabled && !apiKeysProps?.hide && /* @__PURE__ */ jsx(Protect, {
			condition: (has) => has({ permission: "org:sys_api_keys:read" }) || has({ permission: "org:sys_api_keys:manage" }),
			children: /* @__PURE__ */ jsx(Route, {
				path: isAPIKeysPageRoot ? void 0 : "organization-api-keys",
				children: /* @__PURE__ */ jsx(Switch, { children: /* @__PURE__ */ jsx(Route, {
					index: true,
					children: /* @__PURE__ */ jsx(Suspense, {
						fallback: "",
						children: /* @__PURE__ */ jsx(OrganizationAPIKeysPage, {})
					})
				}) })
			})
		})
	] })] });
};

//#endregion
export { OrganizationProfileRoutes };
