// @clerk/ui/dist/components/PaymentMethods/TestPaymentMethod.js
import { localizationKeys } from "../../localization/localizationKeys.js";
import { useLocalizations } from "../../localization/makeLocalizable.js";
import { Box, Text } from "../../customizables/index.js";
import { LineItems } from "../../elements/LineItems.js";
import { jsx, jsxs } from "@emotion/react/jsx-runtime";

//#region src/components/PaymentMethods/TestPaymentMethod.tsx
const TestPaymentMethod = () => {
	const { t } = useLocalizations();
	return /* @__PURE__ */ jsxs(Box, {
		sx: (t$1) => ({
			background: t$1.colors.$neutralAlpha50,
			padding: t$1.space.$2x5,
			borderRadius: t$1.radii.$md,
			borderWidth: t$1.borderWidths.$normal,
			borderStyle: t$1.borderStyles.$solid,
			borderColor: t$1.colors.$borderAlpha100,
			display: "flex",
			flexDirection: "column",
			rowGap: t$1.space.$2,
			position: "relative"
		}),
		children: [
			/* @__PURE__ */ jsx(Box, { sx: (t$1) => ({
				position: "absolute",
				inset: 0,
				background: `repeating-linear-gradient(-45deg,${t$1.colors.$warningAlpha100},${t$1.colors.$warningAlpha100} 6px,${t$1.colors.$warningAlpha150} 6px,${t$1.colors.$warningAlpha150} 12px)`,
				maskImage: `linear-gradient(transparent 20%, black)`,
				pointerEvents: "none"
			}) }),
			/* @__PURE__ */ jsxs(Box, {
				sx: {
					display: "flex",
					alignItems: "baseline",
					justifyContent: "space-between"
				},
				children: [/* @__PURE__ */ jsx(Text, {
					variant: "caption",
					colorScheme: "body",
					localizationKey: localizationKeys("billing.paymentMethod.dev.testCardInfo")
				}), /* @__PURE__ */ jsx(Text, {
					variant: "caption",
					sx: (t$1) => ({
						color: t$1.colors.$warning500,
						fontWeight: t$1.fontWeights.$semibold
					}),
					localizationKey: localizationKeys("billing.paymentMethod.dev.developmentMode")
				})]
			}),
			/* @__PURE__ */ jsxs(LineItems.Root, { children: [
				/* @__PURE__ */ jsxs(LineItems.Group, {
					variant: "tertiary",
					children: [/* @__PURE__ */ jsx(LineItems.Title, { title: localizationKeys("billing.paymentMethod.dev.cardNumber") }), /* @__PURE__ */ jsx(LineItems.Description, { text: "4242 4242 4242 4242" })]
				}),
				/* @__PURE__ */ jsxs(LineItems.Group, {
					variant: "tertiary",
					children: [/* @__PURE__ */ jsx(LineItems.Title, { title: localizationKeys("billing.paymentMethod.dev.expirationDate") }), /* @__PURE__ */ jsx(LineItems.Description, { text: "11/44" })]
				}),
				/* @__PURE__ */ jsxs(LineItems.Group, {
					variant: "tertiary",
					children: [/* @__PURE__ */ jsx(LineItems.Title, { title: localizationKeys("billing.paymentMethod.dev.cvcZip") }), /* @__PURE__ */ jsx(LineItems.Description, { text: t(localizationKeys("billing.paymentMethod.dev.anyNumbers")) })]
				})
			] })
		]
	});
};

//#endregion
export { TestPaymentMethod };
