// @clerk/ui/dist/no-rhc/components/SignUp/index.js
import { SessionTasks } from "../../lazyModules/components.js";
import { Route } from "../../router/Route.js";
import { VIRTUAL_ROUTER_BASE_PATH } from "../../router/VirtualRouter.js";
import { Switch } from "../../router/Switch.js";
import { SignUpContext, useSignUpContext } from "../../contexts/components/SignUp.js";
import { withCoreSessionSwitchGuard } from "../../contexts/CoreSessionContext.js";
import { Flow } from "../../customizables/Flow.js";
import { SignUpEmailLinkFlowComplete } from "../../common/EmailLinkCompleteFlowCard.js";
import { usePreloadTasks } from "../../hooks/usePreloadTasks.js";
import { SignUpStartSolanaWalletsCard } from "./SignUpStartSolanaWalletsCard.js";
import { SignUpContinue } from "./SignUpContinue.js";
import { SignUpEnterpriseConnections } from "./SignUpEnterpriseConnections.js";
import { SignUpSSOCallback } from "./SignUpSSOCallback.js";
import { SignUpStart } from "./SignUpStart.js";
import { SignUpVerifyEmail } from "./SignUpVerifyEmail.js";
import { SignUpVerifyPhone } from "./SignUpVerifyPhone.js";
import React from "react";
import { useClerk } from "@clerk/shared/react";
import { jsx, jsxs } from "@emotion/react/jsx-runtime";

//#region src/components/SignUp/index.tsx
function RedirectToSignUp() {
	const clerk = useClerk();
	React.useEffect(() => {
		clerk.redirectToSignUp();
	}, [clerk]);
	return null;
}
function SignUpRoutes() {
	usePreloadTasks();
	const signUpContext = useSignUpContext();
	return /* @__PURE__ */ jsx(Flow.Root, {
		flow: "signUp",
		children: /* @__PURE__ */ jsxs(Switch, { children: [
			/* @__PURE__ */ jsx(Route, {
				path: "verify-email-address",
				canActivate: (clerk) => !!clerk.client.signUp.emailAddress,
				children: /* @__PURE__ */ jsx(SignUpVerifyEmail, {})
			}),
			/* @__PURE__ */ jsx(Route, {
				path: "verify-phone-number",
				canActivate: (clerk) => !!clerk.client.signUp.phoneNumber,
				children: /* @__PURE__ */ jsx(SignUpVerifyPhone, {})
			}),
			/* @__PURE__ */ jsx(Route, {
				path: "sso-callback",
				children: /* @__PURE__ */ jsx(SignUpSSOCallback, {
					signUpUrl: signUpContext.signUpUrl,
					signInUrl: signUpContext.signInUrl,
					signUpForceRedirectUrl: signUpContext.afterSignUpUrl,
					signInForceRedirectUrl: signUpContext.afterSignInUrl,
					secondFactorUrl: signUpContext.secondFactorUrl,
					continueSignUpUrl: "../continue",
					verifyEmailAddressUrl: "../verify-email-address",
					verifyPhoneNumberUrl: "../verify-phone-number",
					unsafeMetadata: signUpContext.unsafeMetadata
				})
			}),
			/* @__PURE__ */ jsx(Route, {
				path: "verify",
				children: /* @__PURE__ */ jsx(SignUpEmailLinkFlowComplete, {
					redirectUrlComplete: signUpContext.afterSignUpUrl,
					verifyEmailPath: "../verify-email-address",
					verifyPhonePath: "../verify-phone-number"
				})
			}),
			/* @__PURE__ */ jsxs(Route, {
				path: "continue",
				children: [
					/* @__PURE__ */ jsx(Route, {
						path: "verify-email-address",
						canActivate: (clerk) => !!clerk.client.signUp.emailAddress,
						children: /* @__PURE__ */ jsx(SignUpVerifyEmail, {})
					}),
					/* @__PURE__ */ jsx(Route, {
						path: "verify-phone-number",
						canActivate: (clerk) => !!clerk.client.signUp.phoneNumber,
						children: /* @__PURE__ */ jsx(SignUpVerifyPhone, {})
					}),
					/* @__PURE__ */ jsx(Route, {
						index: true,
						children: /* @__PURE__ */ jsx(SignUpContinue, {})
					})
				]
			}),
			/* @__PURE__ */ jsx(Route, {
				path: "tasks",
				children: /* @__PURE__ */ jsx(SessionTasks, { redirectUrlComplete: signUpContext.afterSignUpUrl })
			}),
			/* @__PURE__ */ jsx(Route, {
				path: "enterprise-connections",
				children: /* @__PURE__ */ jsx(SignUpEnterpriseConnections, {})
			}),
			/* @__PURE__ */ jsx(Route, {
				path: "choose-wallet",
				children: /* @__PURE__ */ jsx(SignUpStartSolanaWalletsCard, {})
			}),
			/* @__PURE__ */ jsx(Route, {
				index: true,
				children: /* @__PURE__ */ jsx(SignUpStart, {})
			}),
			/* @__PURE__ */ jsx(Route, { children: /* @__PURE__ */ jsx(RedirectToSignUp, {}) })
		] })
	});
}
SignUpRoutes.displayName = "SignUp";
const SignUp = withCoreSessionSwitchGuard(SignUpRoutes);
const InternalSignUp = withCoreSessionSwitchGuard(SignUpRoutes);
const SignUpModal = (props) => {
	const signUpProps = {
		signInUrl: `/${VIRTUAL_ROUTER_BASE_PATH}/sign-in`,
		waitlistUrl: `/${VIRTUAL_ROUTER_BASE_PATH}/waitlist`,
		...props
	};
	return /* @__PURE__ */ jsx(Route, {
		path: "sign-up",
		children: /* @__PURE__ */ jsx(SignUpContext.Provider, {
			value: {
				componentName: "SignUp",
				...signUpProps,
				routing: "virtual",
				mode: "modal"
			},
			children: /* @__PURE__ */ jsx("div", { children: /* @__PURE__ */ jsx(InternalSignUp, {
				...signUpProps,
				routing: "virtual"
			}) })
		})
	});
};

//#endregion
export { SignUp, SignUpContinue, SignUpModal, SignUpSSOCallback, SignUpStart, SignUpVerifyEmail, SignUpVerifyPhone };
