// @clerk/ui/dist/customizables/index.js
import { localizationKeys } from "../localization/localizationKeys.js";
import { makeLocalizable, useLocalizations } from "../localization/makeLocalizable.js";
import { Box as Box$1 } from "../primitives/Box.js";
import { Col as Col$1, Flex as Flex$1 } from "../primitives/Flex.js";
import { Alert as Alert$1 } from "../primitives/Alert.js";
import { AlertIcon as AlertIcon$1 } from "../primitives/AlertIcon.js";
import { Badge as Badge$1 } from "../primitives/Badge.js";
import { Button as Button$1, SimpleButton as SimpleButton$1 } from "../primitives/Button.js";
import { Dl as Dl$1 } from "../primitives/Dl.js";
import { Dd as Dd$1 } from "../primitives/Dd.js";
import { Dt as Dt$1 } from "../primitives/Dt.js";
import { Form as Form$1 } from "../primitives/Form.js";
import { Text as Text$1 } from "../primitives/Text.js";
import { FormErrorText as FormErrorText$1 } from "../primitives/FormErrorText.js";
import { FormSuccessText as FormSuccessText$1 } from "../primitives/FormSuccessText.js";
import { FormInfoText as FormInfoText$1 } from "../primitives/FormInfoText.js";
import { FormLabel as FormLabel$1 } from "../primitives/FormLabel.js";
import { FormWarningText as FormWarningText$1 } from "../primitives/FormWarningText.js";
import { Grid as Grid$1 } from "../primitives/Grid.js";
import { Heading as Heading$1 } from "../primitives/Heading.js";
import { Icon as Icon$1 } from "../primitives/Icon.js";
import { Image as Image$1 } from "../primitives/Image.js";
import { CheckboxInput as CheckboxInput$1, DivInput, Input as Input$1, RadioInput as RadioInput$1 } from "../primitives/Input.js";
import { Link as Link$1 } from "../primitives/Link.js";
import { NotificationBadge as NotificationBadge$1 } from "../primitives/NotificationBadge.js";
import { Span as Span$1 } from "../primitives/Span.js";
import { Spinner as Spinner$1 } from "../primitives/Spinner.js";
import { Table as Table$1 } from "../primitives/Table.js";
import { Tbody as Tbody$1 } from "../primitives/Tbody.js";
import { Td as Td$1 } from "../primitives/Td.js";
import { Th as Th$1 } from "../primitives/Th.js";
import { Thead as Thead$1 } from "../primitives/Thead.js";
import { Tr as Tr$1 } from "../primitives/Tr.js";
import { Hr as Hr$1 } from "../primitives/Hr.js";
import { descriptors } from "./elementDescriptors.js";
import { AppearanceProvider, useAppearance } from "./AppearanceContext.js";
import { generateFlowPartClassname } from "./classGeneration.js";
import { makeCustomizable } from "./makeCustomizable.js";
import { makeResponsive } from "./makeResponsive.js";
import { sanitizeDomProps } from "./sanitizeDomProps.js";
import { Flow } from "./Flow.js";

//#region src/customizables/index.ts
const Box = makeCustomizable(sanitizeDomProps(Box$1));
const Flex = makeCustomizable(sanitizeDomProps(Flex$1));
const Col = makeCustomizable(sanitizeDomProps(Col$1));
const Grid = makeCustomizable(sanitizeDomProps(Grid$1));
const Button = makeCustomizable(makeLocalizable(sanitizeDomProps(Button$1)), { defaultDescriptor: descriptors.button });
const SimpleButton = makeCustomizable(makeLocalizable(sanitizeDomProps(SimpleButton$1)), { defaultDescriptor: descriptors.button });
const Heading = makeCustomizable(makeLocalizable(sanitizeDomProps(Heading$1)));
const Link = makeCustomizable(makeLocalizable(sanitizeDomProps(Link$1)));
const Text = makeCustomizable(makeLocalizable(sanitizeDomProps(Text$1)));
const Image = makeCustomizable(sanitizeDomProps(makeResponsive(Image$1)));
const Alert = makeCustomizable(sanitizeDomProps(Alert$1));
const AlertIcon = makeCustomizable(sanitizeDomProps(AlertIcon$1));
const Input = makeCustomizable(sanitizeDomProps(Input$1), { defaultDescriptor: descriptors.input });
const OTPInputSegment = makeCustomizable(sanitizeDomProps(DivInput), { defaultDescriptor: descriptors.input });
const CheckboxInput = makeCustomizable(sanitizeDomProps(CheckboxInput$1), { defaultDescriptor: descriptors.checkbox });
const RadioInput = makeCustomizable(sanitizeDomProps(RadioInput$1), { defaultDescriptor: descriptors.radio });
const FormLabel = makeCustomizable(makeLocalizable(sanitizeDomProps(FormLabel$1)));
const FormErrorText = makeCustomizable(makeLocalizable(sanitizeDomProps(FormErrorText$1)));
const FormSuccessText = makeCustomizable(makeLocalizable(sanitizeDomProps(FormSuccessText$1)));
const FormWarningText = makeCustomizable(makeLocalizable(sanitizeDomProps(FormWarningText$1)));
const FormInfoText = makeCustomizable(makeLocalizable(sanitizeDomProps(FormInfoText$1)));
const Form = makeCustomizable(sanitizeDomProps(Form$1));
const Icon = makeCustomizable(sanitizeDomProps(Icon$1));
const Spinner = makeCustomizable(sanitizeDomProps(Spinner$1));
const Badge = makeCustomizable(makeLocalizable(sanitizeDomProps(Badge$1)), { defaultDescriptor: descriptors.badge });
const NotificationBadge = makeCustomizable(makeLocalizable(sanitizeDomProps(NotificationBadge$1)), { defaultDescriptor: descriptors.notificationBadge });
const Table = makeCustomizable(sanitizeDomProps(Table$1), { defaultDescriptor: descriptors.table });
const Thead = makeCustomizable(sanitizeDomProps(Thead$1), { defaultDescriptor: descriptors.tableHead });
const Tbody = makeCustomizable(sanitizeDomProps(Tbody$1), { defaultDescriptor: descriptors.tableBody });
const Tr = makeCustomizable(sanitizeDomProps(Tr$1), { defaultDescriptor: descriptors.tableRow });
const Th = makeCustomizable(makeLocalizable(sanitizeDomProps(Th$1)), { defaultDescriptor: descriptors.tableHeaderCell });
const Td = makeCustomizable(makeLocalizable(sanitizeDomProps(Td$1)), { defaultDescriptor: descriptors.tableBodyCell });
const Dl = makeCustomizable(makeLocalizable(sanitizeDomProps(Dl$1)));
const Dd = makeCustomizable(makeLocalizable(sanitizeDomProps(Dd$1)));
const Dt = makeCustomizable(makeLocalizable(sanitizeDomProps(Dt$1)));
const Span = makeCustomizable(makeLocalizable(sanitizeDomProps(Span$1)));
const Hr = makeCustomizable(sanitizeDomProps(Hr$1));

//#endregion
export { Alert, AlertIcon, AppearanceProvider, Badge, Box, Button, CheckboxInput, Col, Dd, Dl, Dt, Flex, Form, FormErrorText, FormInfoText, FormLabel, FormSuccessText, FormWarningText, Grid, Heading, Icon, Image, Input, Link, NotificationBadge, OTPInputSegment, RadioInput, SimpleButton, Span, Spinner, Table, Tbody, Td, Text, Th, Thead, Tr };
