// @clerk/ui/dist/utils/colors/index.js
import { cssSupports } from "../cssSupports.js";
import { colors } from "./legacy.js";
import { colors as colors$1 } from "./modern.js";

//#region src/utils/colors/index.ts
const colors$2 = {
	changeHslaLightness: colors.changeHslaLightness,
	setHslaAlpha: colors.setHslaAlpha,
	toHslaColor: (color) => {
		if (!color) return;
		return cssSupports.modernColor() ? color : colors.toHslaColor(color);
	},
	toHslaString: (color) => {
		if (!color) return;
		if (cssSupports.modernColor() && typeof color === "string") return color;
		return colors.toHslaString(color);
	},
	lighten: (color, percentage = 0) => {
		if (cssSupports.modernColor()) return colors$1.lighten(color, percentage);
		return colors.lighten(color, percentage);
	},
	makeTransparent: (color, percentage = 0) => {
		if (cssSupports.modernColor()) return colors$1.makeTransparent(color, percentage);
		return colors.makeTransparent(color, percentage);
	},
	makeSolid: (color) => {
		if (cssSupports.modernColor()) return colors$1.makeSolid(color);
		return colors.makeSolid(color);
	},
	setAlpha: (color, alpha) => {
		if (cssSupports.modernColor()) return colors$1.setAlpha(color, alpha);
		return colors.setAlpha(color, alpha);
	},
	adjustForLightness: (color, lightness = 5) => {
		if (cssSupports.modernColor()) return colors$1.adjustForLightness(color, lightness);
		return colors.adjustForLightness(color, lightness);
	}
};

//#endregion
export { colors$2 as colors };
