// @clerk/ui/dist/components/SignIn/index.js
import { SessionTasks } from "../../lazyModules/components.js";
import { useFetch } from "../../hooks/useFetch.js";
import { Route } from "../../router/Route.js";
import { VIRTUAL_ROUTER_BASE_PATH } from "../../router/VirtualRouter.js";
import { Switch } from "../../router/Switch.js";
import { SignInContext, useSignInContext } from "../../contexts/components/SignIn.js";
import { SignUpContext, useSignUpContext } from "../../contexts/components/SignUp.js";
import { withCoreSessionSwitchGuard } from "../../contexts/CoreSessionContext.js";
import { Flow } from "../../customizables/Flow.js";
import { SignInEmailLinkFlowComplete, SignUpEmailLinkFlowComplete } from "../../common/EmailLinkCompleteFlowCard.js";
import { usePreloadTasks } from "../../hooks/usePreloadTasks.js";
import { SignInFactorOneSolanaWalletsCard } from "./SignInFactorOneSolanaWalletsCard.js";
import { normalizeRoutingOptions } from "../../utils/normalizeRoutingOptions.js";
import { LazySignUpContinue, LazySignUpSSOCallback, LazySignUpStart, LazySignUpVerifyEmail, LazySignUpVerifyPhone, preloadSignUp } from "./lazy-sign-up.js";
import { ResetPassword } from "./ResetPassword.js";
import { ResetPasswordSuccess } from "./ResetPasswordSuccess.js";
import { SignInAccountSwitcher } from "./SignInAccountSwitcher.js";
import { SignInClientTrust } from "./SignInClientTrust.js";
import { SignInFactorOne } from "./SignInFactorOne.js";
import { SignInFactorTwo } from "./SignInFactorTwo.js";
import { SignInSSOCallback } from "./SignInSSOCallback.js";
import { SignInStart } from "./SignInStart.js";
import React from "react";
import { useClerk } from "@clerk/shared/react";
import { jsx, jsxs } from "@emotion/react/jsx-runtime";

//#region src/components/SignIn/index.tsx
function RedirectToSignIn() {
	const clerk = useClerk();
	React.useEffect(() => {
		clerk.redirectToSignIn();
	}, [clerk]);
	return null;
}
function SignInRoutes() {
	const signInContext = useSignInContext();
	const signUpContext = useSignUpContext();
	return /* @__PURE__ */ jsx(Flow.Root, {
		flow: "signIn",
		children: /* @__PURE__ */ jsxs(Switch, { children: [
			/* @__PURE__ */ jsx(Route, {
				path: "factor-one",
				children: /* @__PURE__ */ jsx(SignInFactorOne, {})
			}),
			/* @__PURE__ */ jsx(Route, {
				path: "factor-two",
				children: /* @__PURE__ */ jsx(SignInFactorTwo, {})
			}),
			/* @__PURE__ */ jsx(Route, {
				path: "client-trust",
				children: /* @__PURE__ */ jsx(SignInClientTrust, {})
			}),
			/* @__PURE__ */ jsx(Route, {
				path: "reset-password",
				children: /* @__PURE__ */ jsx(ResetPassword, {})
			}),
			/* @__PURE__ */ jsx(Route, {
				path: "reset-password-success",
				children: /* @__PURE__ */ jsx(ResetPasswordSuccess, {})
			}),
			/* @__PURE__ */ jsx(Route, {
				path: "sso-callback",
				children: /* @__PURE__ */ jsx(SignInSSOCallback, {
					signUpUrl: signInContext.signUpUrl,
					signInUrl: signInContext.signInUrl,
					signInForceRedirectUrl: signInContext.afterSignInUrl,
					signUpForceRedirectUrl: signInContext.afterSignUpUrl,
					continueSignUpUrl: signInContext.signUpContinueUrl,
					transferable: signInContext.transferable,
					firstFactorUrl: "../factor-one",
					secondFactorUrl: "../factor-two",
					resetPasswordUrl: "../reset-password",
					unsafeMetadata: signInContext.unsafeMetadata
				})
			}),
			/* @__PURE__ */ jsx(Route, {
				path: "choose",
				children: /* @__PURE__ */ jsx(SignInAccountSwitcher, {})
			}),
			/* @__PURE__ */ jsx(Route, {
				path: "choose-wallet",
				children: /* @__PURE__ */ jsx(SignInFactorOneSolanaWalletsCard, {})
			}),
			/* @__PURE__ */ jsx(Route, {
				path: "verify",
				children: /* @__PURE__ */ jsx(SignInEmailLinkFlowComplete, {
					redirectUrlComplete: signInContext.afterSignInUrl,
					redirectUrl: "../factor-two"
				})
			}),
			signInContext.isCombinedFlow && /* @__PURE__ */ jsxs(Route, {
				path: "create",
				children: [
					/* @__PURE__ */ jsx(Route, {
						path: "verify-email-address",
						canActivate: (clerk) => !!clerk.client.signUp.emailAddress,
						children: /* @__PURE__ */ jsx(LazySignUpVerifyEmail, {})
					}),
					/* @__PURE__ */ jsx(Route, {
						path: "verify-phone-number",
						canActivate: (clerk) => !!clerk.client.signUp.phoneNumber,
						children: /* @__PURE__ */ jsx(LazySignUpVerifyPhone, {})
					}),
					/* @__PURE__ */ jsx(Route, {
						path: "sso-callback",
						children: /* @__PURE__ */ jsx(LazySignUpSSOCallback, {
							signUpUrl: signUpContext.signUpUrl,
							signInUrl: signUpContext.signInUrl,
							signUpForceRedirectUrl: signUpContext.afterSignUpUrl,
							signInForceRedirectUrl: signUpContext.afterSignInUrl,
							secondFactorUrl: signUpContext.secondFactorUrl,
							continueSignUpUrl: "../continue",
							verifyEmailAddressUrl: "../verify-email-address",
							verifyPhoneNumberUrl: "../verify-phone-number",
							unsafeMetadata: signUpContext.unsafeMetadata
						})
					}),
					/* @__PURE__ */ jsx(Route, {
						path: "verify",
						children: /* @__PURE__ */ jsx(SignUpEmailLinkFlowComplete, {
							redirectUrlComplete: signUpContext.afterSignUpUrl,
							verifyEmailPath: "../verify-email-address",
							verifyPhonePath: "../verify-phone-number",
							continuePath: "../continue"
						})
					}),
					/* @__PURE__ */ jsxs(Route, {
						path: "continue",
						children: [
							/* @__PURE__ */ jsx(Route, {
								path: "verify-email-address",
								canActivate: (clerk) => !!clerk.client.signUp.emailAddress,
								children: /* @__PURE__ */ jsx(LazySignUpVerifyEmail, {})
							}),
							/* @__PURE__ */ jsx(Route, {
								path: "verify-phone-number",
								canActivate: (clerk) => !!clerk.client.signUp.phoneNumber,
								children: /* @__PURE__ */ jsx(LazySignUpVerifyPhone, {})
							}),
							/* @__PURE__ */ jsx(Route, {
								index: true,
								children: /* @__PURE__ */ jsx(LazySignUpContinue, {})
							})
						]
					}),
					/* @__PURE__ */ jsx(Route, {
						path: "tasks",
						children: /* @__PURE__ */ jsx(SessionTasks, { redirectUrlComplete: signInContext.afterSignUpUrl })
					}),
					/* @__PURE__ */ jsx(Route, {
						index: true,
						children: /* @__PURE__ */ jsx(LazySignUpStart, {})
					})
				]
			}),
			/* @__PURE__ */ jsx(Route, {
				path: "tasks",
				children: /* @__PURE__ */ jsx(SessionTasks, { redirectUrlComplete: signInContext.afterSignInUrl })
			}),
			/* @__PURE__ */ jsx(Route, {
				index: true,
				children: /* @__PURE__ */ jsx(SignInStart, {})
			}),
			/* @__PURE__ */ jsx(Route, { children: /* @__PURE__ */ jsx(RedirectToSignIn, {}) })
		] })
	});
}
const usePreloadSignUp = (enabled = false) => useFetch(enabled ? preloadSignUp : void 0, "preloadComponent", { staleTime: Infinity });
function SignInRoot() {
	const signInContext = useSignInContext();
	const normalizedSignUpContext = {
		componentName: "SignUp",
		emailLinkRedirectUrl: signInContext.emailLinkRedirectUrl,
		ssoCallbackUrl: signInContext.ssoCallbackUrl,
		forceRedirectUrl: signInContext.signUpForceRedirectUrl,
		fallbackRedirectUrl: signInContext.signUpFallbackRedirectUrl,
		signInUrl: signInContext.signInUrl,
		unsafeMetadata: signInContext.unsafeMetadata,
		...normalizeRoutingOptions({
			routing: signInContext?.routing,
			path: signInContext?.path
		})
	};
	/**
	* Preload Sign Up when in Combined Flow.
	*/
	usePreloadSignUp(signInContext.isCombinedFlow);
	usePreloadTasks();
	return /* @__PURE__ */ jsx(SignUpContext.Provider, {
		value: normalizedSignUpContext,
		children: /* @__PURE__ */ jsx(SignInRoutes, {})
	});
}
SignInRoutes.displayName = "SignIn";
const SignIn = withCoreSessionSwitchGuard(SignInRoot);
const InternalSignIn = withCoreSessionSwitchGuard(SignInRoot);
const SignInModal = (props) => {
	const signInProps = {
		signUpUrl: `/${VIRTUAL_ROUTER_BASE_PATH}/sign-up`,
		waitlistUrl: `/${VIRTUAL_ROUTER_BASE_PATH}/waitlist`,
		...props
	};
	return /* @__PURE__ */ jsx(Route, {
		path: "sign-in",
		children: /* @__PURE__ */ jsx(SignInContext.Provider, {
			value: {
				componentName: "SignIn",
				...signInProps,
				routing: "virtual",
				mode: "modal"
			},
			children: /* @__PURE__ */ jsx("div", { children: /* @__PURE__ */ jsx(InternalSignIn, {
				...signInProps,
				routing: "virtual"
			}) })
		})
	});
};

//#endregion
export { SignIn, SignInModal };
