// @clerk/ui/dist/components/SignIn/SignInFactorOneEnterpriseConnections.js
import { useEnvironment } from "../../contexts/EnvironmentContext.js";
import { localizationKeys } from "../../localization/localizationKeys.js";
import { buildSSOCallbackURL } from "../../common/redirects.js";
import { useSignInContext } from "../../contexts/components/SignIn.js";
import { useCoreSignIn } from "../../contexts/CoreClientContext.js";
import { withCardStateProvider } from "../../elements/contexts/index.js";
import { Flow } from "../../customizables/Flow.js";
import { withRedirect } from "../../common/withRedirect.js";
import { hasMultipleEnterpriseConnections } from "./shared.js";
import { ChooseEnterpriseConnectionCard } from "../../common/ChooseEnterpriseConnectionCard.js";
import { jsx } from "@emotion/react/jsx-runtime";
import { useClerk } from "@clerk/shared/react/index";

//#region src/components/SignIn/SignInFactorOneEnterpriseConnections.tsx
/**
* @experimental
*/
const SignInFactorOneEnterpriseConnectionsInternal = () => {
	const ctx = useSignInContext();
	const { displayConfig } = useEnvironment();
	const signIn = useClerk().client.signIn;
	if (!hasMultipleEnterpriseConnections(signIn.supportedFirstFactors)) return null;
	const enterpriseConnections = signIn.supportedFirstFactors.map((ff) => ({
		id: ff.enterpriseConnectionId,
		name: ff.enterpriseConnectionName
	}));
	const handleEnterpriseSSO = (enterpriseConnectionId) => {
		const redirectUrl = buildSSOCallbackURL(ctx, displayConfig.signInUrl);
		const redirectUrlComplete = ctx.afterSignInUrl || "/";
		return signIn.authenticateWithRedirect({
			strategy: "enterprise_sso",
			redirectUrl,
			redirectUrlComplete,
			oidcPrompt: ctx.oidcPrompt,
			continueSignIn: true,
			enterpriseConnectionId
		});
	};
	return /* @__PURE__ */ jsx(Flow.Part, {
		part: "enterpriseConnections",
		children: /* @__PURE__ */ jsx(ChooseEnterpriseConnectionCard, {
			title: localizationKeys("signIn.enterpriseConnections.title"),
			subtitle: localizationKeys("signIn.enterpriseConnections.subtitle"),
			onClick: handleEnterpriseSSO,
			enterpriseConnections
		})
	});
};
const withEnterpriseConnectionsGuard = (Component) => {
	const displayName = Component.displayName || Component.name || "Component";
	Component.displayName = displayName;
	const HOC = (props) => {
		const signIn = useCoreSignIn();
		const signInCtx = useSignInContext();
		return withRedirect(Component, () => !hasMultipleEnterpriseConnections(signIn.supportedFirstFactors), ({ clerk }) => signInCtx.signInUrl || clerk.buildSignInUrl(), "There are no enterprise connections available to sign-in. Clerk is redirecting to the `signInUrl` instead.")(props);
	};
	HOC.displayName = `withEnterpriseConnectionsGuard(${displayName})`;
	return HOC;
};
const SignInFactorOneEnterpriseConnections = withCardStateProvider(withEnterpriseConnectionsGuard(SignInFactorOneEnterpriseConnectionsInternal));

//#endregion
export { SignInFactorOneEnterpriseConnections };
