// @clerk/ui/dist/components/SignIn/utils.js
import { PREFERRED_SIGN_IN_STRATEGIES } from "../../common/constants.js";
import { getPreferredPhoneCodeChannelByCountry } from "../../utils/phoneUtils.js";
import { otpPrefFactorComparator, passwordPrefFactorComparator } from "../../utils/factorSorting.js";
import { titleize } from "@clerk/shared/underscore";
import { isWebAuthnSupported } from "@clerk/shared/webauthn";

//#region src/components/SignIn/utils.ts
const factorForIdentifier = (i) => (f) => {
	return "safeIdentifier" in f && f.safeIdentifier === i;
};
function findPasskeyStrategy(factors) {
	if (isWebAuthnSupported()) {
		const passkeyFactor = factors.find(({ strategy }) => strategy === "passkey");
		if (passkeyFactor) return passkeyFactor;
	}
	return null;
}
function determineStrategyWhenPasswordIsPreferred(factors, identifier) {
	const passkeyFactor = findPasskeyStrategy(factors);
	if (passkeyFactor) return passkeyFactor;
	const selected = factors.sort(passwordPrefFactorComparator)[0];
	if (selected.strategy === "password") return selected;
	return factors.find(factorForIdentifier(identifier)) || selected || null;
}
function determineStrategyWhenOTPIsPreferred(factors, identifier) {
	const passkeyFactor = findPasskeyStrategy(factors);
	if (passkeyFactor) return passkeyFactor;
	const sortedBasedOnPrefFactor = factors.sort(otpPrefFactorComparator);
	const forIdentifier = sortedBasedOnPrefFactor.find(factorForIdentifier(identifier));
	if (forIdentifier) return forIdentifier;
	const firstBasedOnPref = sortedBasedOnPrefFactor[0];
	if (firstBasedOnPref.strategy === "email_link") return firstBasedOnPref;
	return factors.find(factorForIdentifier(identifier)) || firstBasedOnPref || null;
}
function determineStartingSignInFactor(firstFactors, identifier, preferredSignInStrategy) {
	if (!firstFactors || firstFactors.length === 0) return null;
	return preferredSignInStrategy === PREFERRED_SIGN_IN_STRATEGIES.Password ? determineStrategyWhenPasswordIsPreferred(firstFactors, identifier) : determineStrategyWhenOTPIsPreferred(firstFactors, identifier);
}
const localStrategies = [
	"passkey",
	"email_code",
	"password",
	"phone_code",
	"email_link"
];
function factorHasLocalStrategy(factor) {
	if (!factor) return false;
	return localStrategies.includes(factor.strategy);
}
function determineStartingSignInSecondFactor(secondFactors) {
	if (!secondFactors || secondFactors.length === 0) return null;
	const totpFactor = secondFactors.find((f) => f.strategy === "totp");
	if (totpFactor) return totpFactor;
	const phoneCodeFactor = secondFactors.find((f) => f.strategy === "phone_code");
	if (phoneCodeFactor) return phoneCodeFactor;
	return secondFactors[0];
}
const resetPasswordStrategies = ["reset_password_phone_code", "reset_password_email_code"];
const isResetPasswordStrategy = (strategy) => !!strategy && resetPasswordStrategies.includes(strategy);
const isEmail = (str) => /^\S+@\S+\.\S+$/.test(str);
function getSignUpAttributeFromIdentifier(identifier) {
	if (identifier.type === "tel") return "phoneNumber";
	if (isEmail(identifier.value)) return "emailAddress";
	return "username";
}
const getPreferredAlternativePhoneChannel = (fields, preferredChannels, phoneNumberFieldName) => {
	if (!preferredChannels) return null;
	const strategy = fields.find((f) => f.id === "strategy")?.value;
	if (!!strategy && strategy !== "phone_code") return null;
	const phoneNumber = fields.find((f) => f.id === phoneNumberFieldName)?.value;
	if (!phoneNumber || !phoneNumber?.startsWith("+")) return null;
	const preferredChannel = getPreferredPhoneCodeChannelByCountry(phoneNumber, preferredChannels);
	if (preferredChannel === "sms") return null;
	return preferredChannel;
};
const getPreferredAlternativePhoneChannelForCombinedFlow = (preferredChannels, identifierAttribute, identifierValue) => {
	if (!preferredChannels) return null;
	if (!identifierAttribute || identifierAttribute !== "phoneNumber") return null;
	if (!identifierValue || !identifierValue?.startsWith("+")) return null;
	const preferredChannel = getPreferredPhoneCodeChannelByCountry(identifierValue, preferredChannels);
	if (preferredChannel === "sms") return null;
	return preferredChannel;
};

//#endregion
export { determineStartingSignInFactor, determineStartingSignInSecondFactor, factorHasLocalStrategy, getPreferredAlternativePhoneChannel, getPreferredAlternativePhoneChannelForCombinedFlow, getSignUpAttributeFromIdentifier, isResetPasswordStrategy };
