// @clerk/ui/dist/components/OrganizationSwitcher/index.js
import { withCoreUserGuard } from "../../contexts/CoreUserContext.js";
import { usePopover } from "../../hooks/usePopover.js";
import { useOrganizationSwitcherContext } from "../../contexts/components/OrganizationSwitcher.js";
import { AcceptedInvitationsProvider } from "../../contexts/AcceptedUserInvitations.js";
import { withCardStateProvider, withFloatingTree } from "../../elements/contexts/index.js";
import { Flow } from "../../customizables/Flow.js";
import { Popover } from "../../elements/Popover.js";
import { OrganizationSwitcherPopover } from "./OrganizationSwitcherPopover.js";
import { OrganizationSwitcherTrigger } from "./OrganizationSwitcherTrigger.js";
import { cloneElement, useId } from "react";
import { Fragment as Fragment$1, jsx, jsxs } from "@emotion/react/jsx-runtime";

//#region src/components/OrganizationSwitcher/index.tsx
const OrganizationSwitcherWithFloatingTree = withFloatingTree(({ children }) => {
	const { defaultOpen } = useOrganizationSwitcherContext();
	const { floating, reference, styles, toggle, isOpen, nodeId, context } = usePopover({
		defaultOpen,
		placement: "bottom-start",
		offset: 8
	});
	const switcherButtonMenuId = useId();
	return /* @__PURE__ */ jsxs(Fragment$1, { children: [/* @__PURE__ */ jsx(OrganizationSwitcherTrigger, {
		ref: reference,
		onClick: toggle,
		isOpen,
		"aria-controls": isOpen ? switcherButtonMenuId : void 0,
		"aria-expanded": isOpen
	}), /* @__PURE__ */ jsx(Popover, {
		nodeId,
		context,
		isOpen,
		children: cloneElement(children, {
			id: switcherButtonMenuId,
			close: toggle,
			ref: floating,
			style: styles
		})
	})] });
});
const _OrganizationSwitcher = () => {
	const { __experimental_asStandalone } = useOrganizationSwitcherContext();
	return /* @__PURE__ */ jsx(Flow.Root, {
		flow: "organizationSwitcher",
		sx: { display: "inline-flex" },
		children: /* @__PURE__ */ jsx(AcceptedInvitationsProvider, { children: __experimental_asStandalone ? /* @__PURE__ */ jsx(OrganizationSwitcherPopover, { close: typeof __experimental_asStandalone === "function" ? __experimental_asStandalone : void 0 }) : /* @__PURE__ */ jsx(OrganizationSwitcherWithFloatingTree, { children: /* @__PURE__ */ jsx(OrganizationSwitcherPopover, {}) }) })
	});
};
const OrganizationSwitcher = withCoreUserGuard(withCardStateProvider(_OrganizationSwitcher));

//#endregion
export { OrganizationSwitcher };
