// @clerk/ui/dist/components/UserVerification/UserVerificationFactorOne.js
import { useRouter } from "../../router/RouteContext.js";
import { useEnvironment } from "../../contexts/EnvironmentContext.js";
import { localizationKeys } from "../../localization/localizationKeys.js";
import { useCardState, withCardStateProvider } from "../../elements/contexts/index.js";
import { LoadingCard } from "../../elements/LoadingCard.js";
import { ErrorCard } from "../../elements/ErrorCard.js";
import { determineStartingSignInFactor, factorHasLocalStrategy } from "../SignIn/utils.js";
import { useReverificationAlternativeStrategies } from "./useReverificationAlternativeStrategies.js";
import { useUserVerificationSession, withUserVerificationSessionGuard } from "./useUserVerificationSession.js";
import { AlternativeMethods } from "./AlternativeMethods.js";
import { UserVerificationFactorOnePasswordCard } from "./UserVerificationFactorOnePassword.js";
import { sortByPrimaryFactor } from "./utils.js";
import { UVFactorOneEmailCodeCard } from "./UVFactorOneEmailCodeCard.js";
import { UVFactorOnePasskeysCard } from "./UVFactorOnePasskeysCard.js";
import { UVFactorOnePhoneCodeCard } from "./UVFactorOnePhoneCodeCard.js";
import React, { useEffect, useMemo } from "react";
import { jsx } from "@emotion/react/jsx-runtime";

//#region src/components/UserVerification/UserVerificationFactorOne.tsx
const factorKey = (factor) => {
	if (!factor) return "";
	let key = factor.strategy;
	if ("emailAddressId" in factor) key += factor.emailAddressId;
	if ("phoneNumberId" in factor) key += factor.phoneNumberId;
	return key;
};
const SUPPORTED_STRATEGIES = [
	"password",
	"email_code",
	"phone_code",
	"passkey"
];
function UserVerificationFactorOneInternal() {
	const { data } = useUserVerificationSession();
	const card = useCardState();
	const { navigate } = useRouter();
	const lastPreparedFactorKeyRef = React.useRef("");
	const sessionVerification = data;
	const availableFactors = useMemo(() => {
		return sessionVerification.supportedFirstFactors?.filter((factor) => SUPPORTED_STRATEGIES.includes(factor.strategy))?.sort(sortByPrimaryFactor) || null;
	}, [sessionVerification.supportedFirstFactors]);
	const { preferredSignInStrategy } = useEnvironment().displayConfig;
	const [{ currentFactor }, setFactor] = React.useState(() => ({
		currentFactor: determineStartingSignInFactor(availableFactors, null, preferredSignInStrategy),
		prevCurrentFactor: void 0
	}));
	const { hasAnyStrategy, hasFirstParty } = useReverificationAlternativeStrategies({
		filterOutFactor: currentFactor,
		supportedFirstFactors: availableFactors
	});
	const [showAllStrategies, setShowAllStrategies] = React.useState(() => !currentFactor || !factorHasLocalStrategy(currentFactor));
	const toggleAllStrategies = hasAnyStrategy ? () => {
		card.setError(void 0);
		setShowAllStrategies((s) => !s);
	} : void 0;
	const handleFactorPrepare = () => {
		lastPreparedFactorKeyRef.current = factorKey(currentFactor);
	};
	const selectFactor = (factor) => {
		setFactor((prev) => ({
			currentFactor: factor,
			prevCurrentFactor: prev.currentFactor
		}));
	};
	useEffect(() => {
		if (sessionVerification.status === "needs_second_factor") navigate("factor-two");
	}, []);
	if (!currentFactor) return /* @__PURE__ */ jsx(ErrorCard, {
		cardTitle: localizationKeys("reverification.noAvailableMethods.title"),
		cardSubtitle: localizationKeys("reverification.noAvailableMethods.subtitle"),
		message: localizationKeys("reverification.noAvailableMethods.message"),
		shouldNavigateBack: false
	});
	if (showAllStrategies) {
		const canGoBack = factorHasLocalStrategy(currentFactor);
		const toggle = toggleAllStrategies;
		const backHandler = () => {
			card.setError(void 0);
			toggle?.();
		};
		return /* @__PURE__ */ jsx(AlternativeMethods, {
			onBackLinkClick: canGoBack ? backHandler : void 0,
			onFactorSelected: (f) => {
				selectFactor(f);
				toggle?.();
			},
			currentFactor
		});
	}
	switch (currentFactor?.strategy) {
		case "password": return /* @__PURE__ */ jsx(UserVerificationFactorOnePasswordCard, { onShowAlternativeMethodsClick: toggleAllStrategies });
		case "email_code": return /* @__PURE__ */ jsx(UVFactorOneEmailCodeCard, {
			factorAlreadyPrepared: lastPreparedFactorKeyRef.current === factorKey(currentFactor),
			onFactorPrepare: handleFactorPrepare,
			onShowAlternativeMethodsClicked: toggleAllStrategies,
			factor: currentFactor,
			showAlternativeMethods: hasFirstParty
		});
		case "phone_code": return /* @__PURE__ */ jsx(UVFactorOnePhoneCodeCard, {
			factorAlreadyPrepared: lastPreparedFactorKeyRef.current === factorKey(currentFactor),
			onFactorPrepare: handleFactorPrepare,
			onShowAlternativeMethodsClicked: toggleAllStrategies,
			factor: currentFactor,
			showAlternativeMethods: hasFirstParty
		});
		case "passkey": return /* @__PURE__ */ jsx(UVFactorOnePasskeysCard, { onShowAlternativeMethodsClicked: toggleAllStrategies });
		default: return /* @__PURE__ */ jsx(LoadingCard, {});
	}
}
const UserVerificationFactorOne = withUserVerificationSessionGuard(withCardStateProvider(UserVerificationFactorOneInternal));

//#endregion
export { UserVerificationFactorOne };
