// @clerk/ui/dist/components/UserVerification/UserVerificationFactorTwo.js
import { useRouter } from "../../router/RouteContext.js";
import { withCardStateProvider } from "../../elements/contexts/index.js";
import { LoadingCard } from "../../elements/LoadingCard.js";
import { useSecondFactorSelection } from "../SignIn/useSecondFactorSelection.js";
import { secondFactorsAreEqual } from "./useReverificationAlternativeStrategies.js";
import { useUserVerificationSession, withUserVerificationSessionGuard } from "./useUserVerificationSession.js";
import { sortByPrimaryFactor } from "./utils.js";
import { UserVerificationFactorTwoTOTP } from "./UserVerificationFactorTwoTOTP.js";
import { UVFactorTwoAlternativeMethods } from "./UVFactorTwoAlternativeMethods.js";
import { UVFactorTwoBackupCodeCard } from "./UVFactorTwoBackupCodeCard.js";
import { UVFactorTwoPhoneCodeCard } from "./UVFactorTwoPhoneCodeCard.js";
import { useEffect, useMemo } from "react";
import { jsx } from "@emotion/react/jsx-runtime";

//#region src/components/UserVerification/UserVerificationFactorTwo.tsx
const SUPPORTED_STRATEGIES = [
	"phone_code",
	"totp",
	"backup_code"
];
function UserVerificationFactorTwoComponent() {
	const { navigate } = useRouter();
	const { data } = useUserVerificationSession();
	const sessionVerification = data;
	const availableFactors = useMemo(() => {
		return sessionVerification.supportedSecondFactors?.filter((factor) => SUPPORTED_STRATEGIES.includes(factor.strategy))?.sort(sortByPrimaryFactor) || null;
	}, [sessionVerification.supportedSecondFactors]);
	const { currentFactor, factorAlreadyPrepared, handleFactorPrepare, selectFactor, showAllStrategies, toggleAllStrategies } = useSecondFactorSelection(availableFactors);
	const secondFactorsExcludingCurrent = useMemo(() => availableFactors?.filter((factor) => !secondFactorsAreEqual(factor, currentFactor)), [availableFactors, currentFactor]);
	const hasAlternativeStrategies = useMemo(() => secondFactorsExcludingCurrent && secondFactorsExcludingCurrent.length > 0 || false, [secondFactorsExcludingCurrent]);
	useEffect(() => {
		if (sessionVerification.status === "needs_first_factor") navigate("../");
	}, []);
	if (!currentFactor) return /* @__PURE__ */ jsx(LoadingCard, {});
	if (showAllStrategies && hasAlternativeStrategies) return /* @__PURE__ */ jsx(UVFactorTwoAlternativeMethods, {
		supportedSecondFactors: secondFactorsExcludingCurrent || null,
		onBackLinkClick: toggleAllStrategies,
		onFactorSelected: selectFactor
	});
	switch (currentFactor?.strategy) {
		case "phone_code": return /* @__PURE__ */ jsx(UVFactorTwoPhoneCodeCard, {
			factorAlreadyPrepared,
			onFactorPrepare: handleFactorPrepare,
			factor: currentFactor,
			onShowAlternativeMethodsClicked: toggleAllStrategies,
			showAlternativeMethods: hasAlternativeStrategies
		});
		case "totp": return /* @__PURE__ */ jsx(UserVerificationFactorTwoTOTP, {
			factorAlreadyPrepared,
			onFactorPrepare: handleFactorPrepare,
			factor: currentFactor,
			onShowAlternativeMethodsClicked: toggleAllStrategies,
			showAlternativeMethods: hasAlternativeStrategies
		});
		case "backup_code": return /* @__PURE__ */ jsx(UVFactorTwoBackupCodeCard, { onShowAlternativeMethodsClicked: toggleAllStrategies });
		default: return /* @__PURE__ */ jsx(LoadingCard, {});
	}
}
const UserVerificationFactorTwo = withUserVerificationSessionGuard(withCardStateProvider(UserVerificationFactorTwoComponent));

//#endregion
export { UserVerificationFactorTwo };
