// @clerk/ui/dist/elements/Card/CardRoot.js
import { mqu } from "../../styledSystem/breakpoints.js";
import { descriptors } from "../../customizables/elementDescriptors.js";
import { useAppearance } from "../../customizables/AppearanceContext.js";
import { generateFlowPartClassname } from "../../customizables/classGeneration.js";
import { useFlowMetadata } from "../contexts/index.js";
import { Col } from "../../customizables/index.js";
import { ApplicationLogo } from "../ApplicationLogo.js";
import React from "react";
import { Fragment as Fragment$1, jsx, jsxs } from "@emotion/react/jsx-runtime";

//#region src/elements/Card/CardRoot.tsx
const CardRoot = React.forwardRef((props, ref) => {
	const { sx, children,...rest } = props;
	const appearance = useAppearance();
	const flowMetadata = useFlowMetadata();
	return /* @__PURE__ */ jsxs(Fragment$1, { children: [appearance.parsedOptions.logoPlacement === "outside" && /* @__PURE__ */ jsx(ApplicationLogo, { sx: (t) => ({
		position: "relative",
		[mqu.sm]: { margin: `0 0 ${t.space.$7} 0` }
	}) }), /* @__PURE__ */ jsx(Col, {
		elementDescriptor: [descriptors.cardBox, props.elementDescriptor],
		className: generateFlowPartClassname(flowMetadata),
		ref,
		sx: [(t) => ({
			isolation: "isolate",
			maxWidth: `calc(100vw - ${t.sizes.$10})`,
			width: t.sizes.$100,
			borderWidth: t.borderWidths.$normal,
			borderStyle: t.borderStyles.$solid,
			borderColor: t.colors.$borderAlpha150,
			borderRadius: t.radii.$xl,
			color: t.colors.$colorForeground,
			position: "relative",
			overflow: "hidden"
		}), sx],
		...rest,
		children
	})] });
});

//#endregion
export { CardRoot };
