// @clerk/ui/dist/no-rhc/primitives/Input.js
import { localizationKeys } from "../localization/localizationKeys.js";
import { useLocalizations } from "../localization/makeLocalizable.js";
import { createVariants } from "../styledSystem/createVariants.js";
import { common } from "../styledSystem/common.js";
import { mqu } from "../styledSystem/breakpoints.js";
import { sanitizeInputProps, useFormField } from "./hooks/useFormField.js";
import { useInput } from "./hooks/useInput.js";
import React from "react";
import { jsx } from "@emotion/react/jsx-runtime";

//#region src/primitives/Input.tsx
const { applyVariants, filterProps } = createVariants((theme, props) => ({
	base: {
		boxSizing: "border-box",
		margin: 0,
		padding: `${theme.space.$1x5} ${theme.space.$3}`,
		backgroundColor: theme.colors.$colorInput,
		color: theme.colors.$colorInputForeground,
		outline: "transparent solid 2px",
		outlineOffset: "2px",
		maxHeight: theme.sizes.$9,
		width: props.type === "checkbox" ? theme.sizes.$4 : "100%",
		aspectRatio: props.type === "checkbox" ? "1/1" : "unset",
		accentColor: theme.colors.$primary500,
		...common.textVariants(theme).body,
		...common.disabled(theme),
		[mqu.ios]: {
			fontSize: theme.fontSizes.$lg,
			"&:not([type=\"checkbox\"]):not([type=\"radio\"])": { WebkitAppearance: "none" }
		},
		":autofill": { animationName: "onAutoFillStart" },
		"::placeholder": { color: theme.colors.$colorMutedForeground }
	},
	variants: { variant: {
		default: { ...common.borderVariants(theme, props).normal },
		unstyled: {
			borderWidth: 0,
			boxShadow: "unset",
			backgroundColor: "transparent"
		}
	} },
	defaultVariants: { variant: "default" }
}));
const Input = React.forwardRef((props, ref) => {
	const fieldControl = useFormField() || {};
	const { t } = useLocalizations();
	const { feedbackMessageId, ignorePasswordManager, feedbackType,...fieldControlProps } = sanitizeInputProps(fieldControl, [
		"feedbackMessageId",
		"ignorePasswordManager",
		"feedbackType"
	]);
	const propsWithoutVariants = filterProps({
		...props,
		hasError: props.hasError || fieldControlProps.hasError
	});
	const { onChange } = useInput(propsWithoutVariants.onChange);
	const { isDisabled, hasError, focusRing, isRequired, type,...rest } = propsWithoutVariants;
	const _disabled = isDisabled || fieldControlProps.isDisabled;
	const _required = isRequired || fieldControlProps.isRequired;
	const _hasError = hasError || fieldControlProps.hasError;
	/**
	* type="email" will not allow characters like this one "ö", instead remove type email and provide a pattern that accepts any character before the "@" symbol
	* inputMode="email" ensures the email keyboard appears on mobile devices
	*/
	const typeProps = type === "email" ? {
		type: "text",
		pattern: "^.*@[a-zA-Z0-9\\-]+\\.[a-zA-Z0-9\\-\\.]+$",
		inputMode: "email",
		title: t(localizationKeys("formFieldInput__emailAddress_format"))
	} : { type: type || "text" };
	const passwordManagerProps = ignorePasswordManager ? { "data-1p-ignore": true } : void 0;
	return /* @__PURE__ */ jsx("input", {
		...rest,
		...typeProps,
		...passwordManagerProps,
		ref,
		onChange,
		disabled: isDisabled,
		required: _required,
		id: props.id || fieldControlProps.id,
		"aria-invalid": _hasError,
		"aria-describedby": feedbackMessageId || void 0,
		"aria-required": _required,
		"aria-disabled": _disabled,
		"data-feedback": feedbackType,
		"data-variant": props.variant || "default",
		css: applyVariants(props)
	});
});
const DivInput = React.forwardRef((props, ref) => {
	const { feedbackType } = sanitizeInputProps(useFormField() || {}, ["feedbackType", "data-active"]);
	const { isDisabled, hasError, focusRing, isRequired,...rest } = filterProps({
		...props,
		hasError: props.hasError
	});
	return /* @__PURE__ */ jsx("div", {
		...rest,
		ref,
		id: props.id,
		"aria-invalid": hasError,
		"data-feedback": feedbackType,
		"data-variant": props.variant || "default",
		"data-focus-within": focusRing,
		css: applyVariants(props)
	});
});
const CheckboxInput = React.forwardRef((props, ref) => {
	return /* @__PURE__ */ jsx(Input, {
		...props,
		type: "checkbox",
		ref
	});
});
const RadioInput = React.forwardRef((props, ref) => {
	return /* @__PURE__ */ jsx(Input, {
		...props,
		type: "radio",
		ref
	});
});
const Textarea = React.forwardRef((props, ref) => {
	const { isDisabled, hasError, focusRing, isRequired,...rest } = props;
	const _disabled = isDisabled || props.isDisabled;
	const _required = isRequired || props.isRequired;
	const _hasError = hasError || props.hasError;
	return /* @__PURE__ */ jsx("textarea", {
		...rest,
		ref,
		disabled: _disabled,
		required: _required,
		"aria-invalid": _hasError,
		"aria-required": _required,
		css: applyVariants(props)
	});
});

//#endregion
export { CheckboxInput, DivInput, Input, RadioInput, Textarea };
