// @clerk/ui/dist/components/UserProfile/EnterpriseAccountsSection.js
import { localizationKeys } from "../../localization/localizationKeys.js";
import { descriptors } from "../../customizables/elementDescriptors.js";
import { Badge, Box, Flex, Text } from "../../customizables/index.js";
import { ProviderIcon } from "../../common/ProviderIcon.js";
import { ProfileSection } from "../../elements/Section.js";
import { useUser } from "@clerk/shared/react";
import { jsx, jsxs } from "@emotion/react/jsx-runtime";

//#region src/components/UserProfile/EnterpriseAccountsSection.tsx
const EnterpriseAccountsSection = () => {
	const { user } = useUser();
	const activeEnterpriseAccounts = user?.enterpriseAccounts.filter(({ enterpriseConnection }) => enterpriseConnection?.active);
	if (!activeEnterpriseAccounts?.length) return null;
	return /* @__PURE__ */ jsx(ProfileSection.Root, {
		title: localizationKeys("userProfile.start.enterpriseAccountsSection.title"),
		id: "enterpriseAccounts",
		centered: false,
		children: /* @__PURE__ */ jsx(ProfileSection.ItemList, {
			id: "enterpriseAccounts",
			children: activeEnterpriseAccounts.map((account) => /* @__PURE__ */ jsx(EnterpriseAccount, { account }, account.id))
		})
	});
};
const EnterpriseAccount = ({ account }) => {
	const label = account.emailAddress;
	const connectionName = account?.enterpriseConnection?.name;
	const error = account.verification?.error?.longMessage;
	return /* @__PURE__ */ jsxs(ProfileSection.Item, {
		id: "enterpriseAccounts",
		sx: (t) => ({
			gap: t.space.$2,
			justifyContent: "start"
		}),
		children: [/* @__PURE__ */ jsx(EnterpriseAccountProviderIcon, { account }), /* @__PURE__ */ jsx(Box, {
			sx: {
				whiteSpace: "nowrap",
				overflow: "hidden"
			},
			children: /* @__PURE__ */ jsxs(Flex, {
				gap: 1,
				center: true,
				children: [
					/* @__PURE__ */ jsx(Text, {
						truncate: true,
						colorScheme: "body",
						children: connectionName
					}),
					/* @__PURE__ */ jsx(Text, {
						truncate: true,
						as: "span",
						colorScheme: "secondary",
						children: label ? `• ${label}` : ""
					}),
					error && /* @__PURE__ */ jsx(Badge, {
						colorScheme: "danger",
						localizationKey: localizationKeys("badge__requiresAction")
					})
				]
			})
		})]
	}, account.id);
};
const EnterpriseAccountProviderIcon = ({ account }) => {
	const { provider, enterpriseConnection } = account;
	const providerWithoutPrefix = provider.replace(/(oauth_|saml_)/, "").trim();
	const connectionName = enterpriseConnection?.name ?? providerWithoutPrefix;
	return /* @__PURE__ */ jsx(ProviderIcon, {
		id: providerWithoutPrefix,
		iconUrl: enterpriseConnection?.logoPublicUrl,
		name: connectionName,
		alt: `${connectionName}'s icon`,
		elementDescriptor: [descriptors.providerIcon],
		elementId: descriptors.enterpriseButtonsProviderIcon.setId(account.provider)
	});
};

//#endregion
export { EnterpriseAccountsSection };
