// @clerk/ui/dist/components/UserProfile/Web3Form.js
import { useEnabledThirdPartyProviders } from "../../hooks/useEnabledThirdPartyProviders.js";
import { useModuleManager } from "../../contexts/ModuleManagerContext.js";
import { localizationKeys } from "../../localization/localizationKeys.js";
import { descriptors } from "../../customizables/elementDescriptors.js";
import { useCardState } from "../../elements/contexts/index.js";
import { ProviderIcon } from "../../common/ProviderIcon.js";
import { useActionContext } from "../../elements/Action/ActionRoot.js";
import { Action } from "../../elements/Action/index.js";
import { ProfileSection } from "../../elements/Section.js";
import { web3CallbackErrorHandler } from "../../utils/web3CallbackErrorHandler.js";
import { Web3SelectSolanaWalletScreen } from "./Web3SelectSolanaWalletScreen.js";
import { useReverification, useUser } from "@clerk/shared/react";
import { Fragment, jsx, jsxs } from "@emotion/react/jsx-runtime";
import { createWeb3 } from "@clerk/shared/internal/clerk-js/web3";

//#region src/components/UserProfile/Web3Form.tsx
const AddWeb3WalletActionMenu = () => {
	const card = useCardState();
	const { open } = useActionContext();
	const { user } = useUser();
	const { strategies, strategyToDisplayData } = useEnabledThirdPartyProviders();
	const moduleManager = useModuleManager();
	const enabledStrategies = strategies.filter((s) => s.startsWith("web3"));
	const connectedStrategies = user?.verifiedWeb3Wallets?.map((w) => w.verification.strategy) ?? [];
	const unconnectedStrategies = enabledStrategies.filter((strategy) => {
		return !connectedStrategies.includes(strategy) && strategyToDisplayData[strategy];
	});
	if (unconnectedStrategies.length === 0) return null;
	const createWeb3Wallet = useReverification((identifier) => user?.createWeb3Wallet({ web3Wallet: identifier }));
	const connect = async ({ strategy, walletName }) => {
		const web3 = createWeb3(moduleManager);
		if (strategy === "web3_solana_signature" && !walletName) {
			open("web3Wallets");
			return;
		}
		const provider = strategy.replace("web3_", "").replace("_signature", "");
		card.setError(void 0);
		try {
			card.setLoading(strategy);
			const identifier = await web3.getWeb3Identifier({
				provider,
				walletName
			});
			if (!user) throw new Error("user is not defined");
			let web3Wallet = await createWeb3Wallet(identifier);
			web3Wallet = await web3Wallet?.prepareVerification({ strategy });
			const message = web3Wallet?.verification.message;
			const signature = await web3.generateWeb3Signature({
				identifier,
				nonce: message,
				provider,
				walletName
			});
			await web3Wallet?.attemptVerification({ signature });
			card.setIdle();
		} catch (err) {
			card.setIdle();
			web3CallbackErrorHandler(err, card.setError);
		}
	};
	return /* @__PURE__ */ jsxs(Fragment, { children: [/* @__PURE__ */ jsx(Action.Closed, {
		value: "web3Wallets",
		children: /* @__PURE__ */ jsx(ProfileSection.ActionMenu, {
			id: "web3Wallets",
			triggerLocalizationKey: localizationKeys("userProfile.start.web3WalletsSection.primaryButton"),
			children: unconnectedStrategies.map((strategy) => /* @__PURE__ */ jsx(ProfileSection.ActionMenuItem, {
				id: strategyToDisplayData[strategy].id,
				onClick: () => {
					connect({ strategy });
				},
				isLoading: card.loadingMetadata === strategy,
				isDisabled: card.isLoading,
				localizationKey: localizationKeys("userProfile.web3WalletPage.web3WalletButtonsBlockButton", { provider: strategyToDisplayData[strategy].name }),
				sx: (t) => ({
					justifyContent: "start",
					gap: t.space.$2
				}),
				leftIcon: /* @__PURE__ */ jsx(ProviderIcon, {
					id: strategyToDisplayData[strategy].id,
					iconUrl: strategyToDisplayData[strategy].iconUrl,
					name: strategyToDisplayData[strategy].name,
					isLoading: card.loadingMetadata === strategy,
					isDisabled: card.isLoading,
					alt: `Connect ${strategyToDisplayData[strategy].name}`,
					elementDescriptor: descriptors.providerIcon,
					elementId: descriptors.providerIcon.setId(strategyToDisplayData[strategy].id)
				})
			}, strategy))
		})
	}), /* @__PURE__ */ jsx(Action.Open, {
		value: "web3Wallets",
		children: /* @__PURE__ */ jsx(Action.Card, { children: /* @__PURE__ */ jsx(Web3SelectSolanaWalletScreen, { onConnect: connect }) })
	})] });
};

//#endregion
export { AddWeb3WalletActionMenu };
