// @clerk/ui/dist/no-rhc/components/UserVerification/AlternativeMethods.js
import { localizationKeys } from "../../localization/localizationKeys.js";
import chat_alt_default from "../../icons/chat-alt.js";
import email_default from "../../icons/email.js";
import fingerprint_default from "../../icons/fingerprint.js";
import lock_closed_default from "../../icons/lock-closed.js";
import organization_default from "../../icons/organization.js";
import { descriptors } from "../../customizables/elementDescriptors.js";
import { useCardState } from "../../elements/contexts/index.js";
import { Flow } from "../../customizables/Flow.js";
import { Col, Flex } from "../../customizables/index.js";
import { Card } from "../../elements/Card/index.js";
import { Header } from "../../elements/Header.js";
import { formatSafeIdentifier } from "../../utils/formatSafeIdentifier.js";
import { ArrowBlockButton } from "../../elements/ArrowBlockButton.js";
import { BackLink } from "../../elements/BackLink.js";
import { useReverificationAlternativeStrategies } from "./useReverificationAlternativeStrategies.js";
import { useUserVerificationSession } from "./useUserVerificationSession.js";
import { withHavingTrouble } from "./withHavingTrouble.js";
import React from "react";
import { jsx, jsxs } from "@emotion/react/jsx-runtime";

//#region src/components/UserVerification/AlternativeMethods.tsx
const AlternativeMethods = (props) => {
	return withHavingTrouble(AlternativeMethodsList, { ...props });
};
const AlternativeMethodsList = (props) => {
	const { onBackLinkClick, onHavingTroubleClick, onFactorSelected } = props;
	const card = useCardState();
	const { data } = useUserVerificationSession();
	const { firstPartyFactors, hasAnyStrategy } = useReverificationAlternativeStrategies({
		filterOutFactor: props?.currentFactor,
		supportedFirstFactors: data?.supportedFirstFactors
	});
	return /* @__PURE__ */ jsx(Flow.Part, {
		part: "alternativeMethods",
		children: /* @__PURE__ */ jsxs(Card.Root, { children: [/* @__PURE__ */ jsxs(Card.Content, { children: [
			/* @__PURE__ */ jsxs(Header.Root, { children: [/* @__PURE__ */ jsx(Header.Title, { localizationKey: localizationKeys("reverification.alternativeMethods.title") }), /* @__PURE__ */ jsx(Header.Subtitle, { localizationKey: localizationKeys("reverification.alternativeMethods.subtitle") })] }),
			/* @__PURE__ */ jsx(Card.Alert, { children: card.error }),
			/* @__PURE__ */ jsx(Flex, {
				direction: "col",
				elementDescriptor: descriptors.main,
				gap: 6,
				children: /* @__PURE__ */ jsxs(Col, {
					gap: 4,
					children: [hasAnyStrategy && /* @__PURE__ */ jsx(Flex, {
						elementDescriptor: descriptors.alternativeMethods,
						direction: "col",
						gap: 2,
						children: firstPartyFactors.map((factor, i) => /* @__PURE__ */ jsx(ArrowBlockButton, {
							leftIcon: getButtonIcon(factor),
							textLocalizationKey: getButtonLabel(factor),
							elementDescriptor: descriptors.alternativeMethodsBlockButton,
							textElementDescriptor: descriptors.alternativeMethodsBlockButtonText,
							arrowElementDescriptor: descriptors.alternativeMethodsBlockButtonArrow,
							textVariant: "buttonLarge",
							isDisabled: card.isLoading,
							onClick: () => {
								card.setError(void 0);
								onFactorSelected(factor);
							}
						}, i))
					}), onBackLinkClick && /* @__PURE__ */ jsx(BackLink, {
						boxElementDescriptor: descriptors.backRow,
						linkElementDescriptor: descriptors.backLink,
						onClick: onBackLinkClick
					})]
				})
			})
		] }), /* @__PURE__ */ jsx(Card.Footer, { children: /* @__PURE__ */ jsxs(Card.Action, {
			elementId: "havingTrouble",
			children: [/* @__PURE__ */ jsx(Card.ActionText, { localizationKey: localizationKeys("reverification.alternativeMethods.actionText") }), /* @__PURE__ */ jsx(Card.ActionLink, {
				localizationKey: localizationKeys("reverification.alternativeMethods.actionLink"),
				onClick: onHavingTroubleClick
			})]
		}) })] })
	});
};
function getButtonLabel(factor) {
	switch (factor.strategy) {
		case "email_code": return localizationKeys("reverification.alternativeMethods.blockButton__emailCode", { identifier: formatSafeIdentifier(factor.safeIdentifier) || "" });
		case "phone_code": return localizationKeys("reverification.alternativeMethods.blockButton__phoneCode", { identifier: formatSafeIdentifier(factor.safeIdentifier) || "" });
		case "password": return localizationKeys("reverification.alternativeMethods.blockButton__password");
		case "passkey": return localizationKeys("reverification.alternativeMethods.blockButton__passkey");
		default: throw new Error(`Invalid sign in strategy: "${factor.strategy}"`);
	}
}
function getButtonIcon(factor) {
	return {
		email_code: email_default,
		phone_code: chat_alt_default,
		password: lock_closed_default,
		passkey: fingerprint_default,
		enterprise_sso: organization_default
	}[factor.strategy];
}

//#endregion
export { AlternativeMethods };
