// @clerk/ui/dist/no-rhc/lazyModules/components.js
import { lazy } from "react";

//#region src/lazyModules/components.ts
const componentImportPaths = {
	SignIn: () => import(
		/* webpackChunkName: "signin" */
		"../components/SignIn/index.js"
),
	SignUp: () => import(
		/* webpackChunkName: "signup" */
		"../components/SignUp/index.js"
),
	UserAvatar: () => import(
		/* webpackChunkName: "useravatar" */
		"../components/UserAvatar/index.js"
),
	UserButton: () => import(
		/* webpackChunkName: "userbutton" */
		"../components/UserButton/index.js"
),
	UserProfile: () => import(
		/* webpackChunkName: "userprofile" */
		"../components/UserProfile/index.js"
),
	CreateOrganization: () => import(
		/* webpackChunkName: "createorganization" */
		"../components/CreateOrganization/index.js"
),
	OrganizationProfile: () => import(
		/* webpackChunkName: "organizationprofile" */
		"../components/OrganizationProfile/index.js"
),
	OrganizationSwitcher: () => import(
		/* webpackChunkName: "organizationswitcher" */
		"../components/OrganizationSwitcher/index.js"
),
	OrganizationList: () => import(
		/* webpackChunkName: "organizationlist" */
		"../components/OrganizationList/index.js"
),
	ImpersonationFab: () => import(
		/* webpackChunkName: "impersonationfab" */
		"../components/ImpersonationFab/index.js"
),
	GoogleOneTap: () => import(
		/* webpackChunkName: "onetap" */
		"../components/GoogleOneTap/index.js"
),
	BlankCaptchaModal: () => import(
		/* webpackChunkName: "blankcaptcha" */
		"../components/BlankCaptchaModal/index.js"
),
	UserVerification: () => import(
		/* webpackChunkName: "userverification" */
		"../components/UserVerification/index.js"
),
	Waitlist: () => import(
		/* webpackChunkName: "waitlist" */
		"../components/Waitlist/index.js"
),
	KeylessPrompt: () => import(
		/* webpackChunkName: "keylessPrompt" */
		"../components/devPrompts/KeylessPrompt/index.js"
),
	PricingTable: () => import(
		/* webpackChunkName: "pricingTable" */
		"../components/PricingTable/index.js"
),
	Checkout: () => import(
		/* webpackChunkName: "checkout" */
		"../components/Checkout/index.js"
),
	SessionTasks: () => import(
		/* webpackChunkName: "sessionTasks" */
		"../components/SessionTasks/index.js"
),
	TaskChooseOrganization: () => import(
		/* webpackChunkName: "taskChooseOrganization" */
		"../components/SessionTasks/tasks/TaskChooseOrganization/index.js"
),
	TaskResetPassword: () => import(
		/* webpackChunkName: "taskResetPassword" */
		"../components/SessionTasks/tasks/TaskResetPassword/index.js"
),
	TaskSetupMFA: () => import(
		/* webpackChunkName: "taskSetupMFA" */
		"../components/SessionTasks/tasks/TaskSetupMfa/index.js"
),
	PlanDetails: () => import(
		/* webpackChunkName: "planDetails" */
		"../components/Plans/PlanDetails.js"
),
	SubscriptionDetails: () => import(
		/* webpackChunkName: "subscriptionDetails" */
		"../components/SubscriptionDetails/index.js"
),
	APIKeys: () => import(
		/* webpackChunkName: "apiKeys" */
		"../components/APIKeys/APIKeys.js"
),
	OAuthConsent: () => import(
		/* webpackChunkName: "oauthConsent" */
		"../components/OAuthConsent/OAuthConsent.js"
),
	EnableOrganizationsPrompt: () => import(
		/* webpackChunkName: "enableOrganizationsPrompt" */
		"../components/devPrompts/EnableOrganizationsPrompt/index.js"
)
};
const SignIn = lazy(() => componentImportPaths.SignIn().then((module) => ({ default: module.SignIn })));
const SignInModal = lazy(() => componentImportPaths.SignIn().then((module) => ({ default: module.SignInModal })));
const GoogleOneTap = lazy(() => componentImportPaths.GoogleOneTap().then((module) => ({ default: module.OneTap })));
const UserVerification = lazy(() => componentImportPaths.UserVerification().then((module) => ({ default: module.UserVerification })));
const UserVerificationModal = lazy(() => componentImportPaths.UserVerification().then((module) => ({ default: module.UserVerificationModal })));
const EnableOrganizationsPrompt = lazy(() => componentImportPaths.EnableOrganizationsPrompt().then((module) => ({ default: module.EnableOrganizationsPrompt })));
const SignUp = lazy(() => componentImportPaths.SignUp().then((module) => ({ default: module.SignUp })));
const SignUpModal = lazy(() => componentImportPaths.SignUp().then((module) => ({ default: module.SignUpModal })));
const UserAvatar = lazy(() => componentImportPaths.UserAvatar().then((module) => ({ default: module.UserAvatar })));
const UserButton = lazy(() => componentImportPaths.UserButton().then((module) => ({ default: module.UserButton })));
const UserProfile = lazy(() => componentImportPaths.UserProfile().then((module) => ({ default: module.UserProfile })));
const UserProfileModal = lazy(() => componentImportPaths.UserProfile().then((module) => ({ default: module.UserProfileModal })));
const CreateOrganization = lazy(() => componentImportPaths.CreateOrganization().then((module) => ({ default: module.CreateOrganization })));
const CreateOrganizationModal = lazy(() => componentImportPaths.CreateOrganization().then((module) => ({ default: module.CreateOrganizationModal })));
const OrganizationProfile = lazy(() => componentImportPaths.OrganizationProfile().then((module) => ({ default: module.OrganizationProfile })));
const OrganizationProfileModal = lazy(() => componentImportPaths.OrganizationProfile().then((module) => ({ default: module.OrganizationProfileModal })));
const OrganizationSwitcher = lazy(() => componentImportPaths.OrganizationSwitcher().then((module) => ({ default: module.OrganizationSwitcher })));
const OrganizationList = lazy(() => componentImportPaths.OrganizationList().then((module) => ({ default: module.OrganizationList })));
const Waitlist = lazy(() => componentImportPaths.Waitlist().then((module) => ({ default: module.Waitlist })));
const WaitlistModal = lazy(() => componentImportPaths.Waitlist().then((module) => ({ default: module.WaitlistModal })));
const BlankCaptchaModal = lazy(() => componentImportPaths.BlankCaptchaModal().then((module) => ({ default: module.BlankCaptchaModal })));
const ImpersonationFab = lazy(() => componentImportPaths.ImpersonationFab().then((module) => ({ default: module.ImpersonationFab })));
const KeylessPrompt = lazy(() => componentImportPaths.KeylessPrompt().then((module) => ({ default: module.KeylessPrompt })));
const PricingTable = lazy(() => componentImportPaths.PricingTable().then((module) => ({ default: module.PricingTable })));
const APIKeys = lazy(() => componentImportPaths.APIKeys().then((module) => ({ default: module.APIKeys })));
const Checkout = lazy(() => componentImportPaths.Checkout().then((module) => ({ default: module.Checkout })));
const TaskChooseOrganization = lazy(() => componentImportPaths.TaskChooseOrganization().then((module) => ({ default: module.TaskChooseOrganization })));
const TaskResetPassword = lazy(() => componentImportPaths.TaskResetPassword().then((module) => ({ default: module.TaskResetPassword })));
const TaskSetupMFA = lazy(() => componentImportPaths.TaskSetupMFA().then((module) => ({ default: module.TaskSetupMFA })));
const PlanDetails = lazy(() => componentImportPaths.PlanDetails().then((module) => ({ default: module.PlanDetails })));
const SubscriptionDetails = lazy(() => componentImportPaths.SubscriptionDetails().then((module) => ({ default: module.SubscriptionDetails })));
const OAuthConsent = lazy(() => componentImportPaths.OAuthConsent().then((module) => ({ default: module.OAuthConsent })));
const SessionTasks = lazy(() => componentImportPaths.SessionTasks().then((module) => ({ default: module.SessionTasks })));
const preloadComponent = async (component) => {
	return componentImportPaths[component]?.();
};
const ClerkComponents = {
	SignIn,
	SignUp,
	UserAvatar,
	UserButton,
	UserProfile,
	UserVerification,
	EnableOrganizationsPrompt,
	OrganizationSwitcher,
	OrganizationList,
	OrganizationProfile,
	CreateOrganization,
	SignInModal,
	SignUpModal,
	UserProfileModal,
	OrganizationProfileModal,
	CreateOrganizationModal,
	UserVerificationModal,
	GoogleOneTap,
	Waitlist,
	WaitlistModal,
	BlankCaptchaModal,
	PricingTable,
	Checkout,
	PlanDetails,
	APIKeys,
	OAuthConsent,
	SubscriptionDetails,
	TaskChooseOrganization,
	TaskResetPassword,
	TaskSetupMFA
};

//#endregion
export { BlankCaptchaModal, Checkout, ClerkComponents, CreateOrganizationModal, EnableOrganizationsPrompt, ImpersonationFab, KeylessPrompt, OrganizationProfileModal, PlanDetails, SessionTasks, SignInModal, SignUpModal, UserProfileModal, UserVerificationModal, WaitlistModal, preloadComponent };
