// @clerk/ui/dist/no-rhc/components/UserProfile/MfaTOTPScreen.js
import { localizationKeys } from "../../localization/localizationKeys.js";
import { withCardStateProvider } from "../../elements/contexts/index.js";
import { Wizard, useWizard } from "../../common/Wizard.js";
import { SuccessPage } from "../../elements/SuccessPage.js";
import { MfaBackupCodeList } from "./MfaBackupCodeList.js";
import { AddAuthenticatorApp } from "./AddAuthenticatorApp.js";
import { VerifyTOTP } from "./VerifyTOTP.js";
import React from "react";
import { jsx, jsxs } from "@emotion/react/jsx-runtime";

//#region src/components/UserProfile/MfaTOTPScreen.tsx
const MfaTOTPScreen = withCardStateProvider((props) => {
	const { onReset } = props;
	const wizard = useWizard();
	const ref = React.useRef();
	return /* @__PURE__ */ jsxs(Wizard, {
		...wizard.props,
		children: [
			/* @__PURE__ */ jsx(AddAuthenticatorApp, {
				title: localizationKeys("userProfile.mfaTOTPPage.title"),
				onSuccess: wizard.nextStep,
				onReset
			}),
			/* @__PURE__ */ jsx(VerifyTOTP, {
				onSuccess: wizard.nextStep,
				onReset,
				resourceRef: ref
			}),
			/* @__PURE__ */ jsx(SuccessPage, {
				title: localizationKeys("userProfile.mfaTOTPPage.title"),
				text: localizationKeys("userProfile.mfaTOTPPage.successMessage"),
				onFinish: onReset,
				contents: /* @__PURE__ */ jsx(MfaBackupCodeList, {
					subtitle: localizationKeys("userProfile.backupCodePage.successSubtitle"),
					backupCodes: ref.current?.backupCodes
				})
			})
		]
	});
});

//#endregion
export { MfaTOTPScreen };
