// @clerk/ui/dist/no-rhc/components/UserProfile/PasskeySection.js
import { localizationKeys } from "../../localization/localizationKeys.js";
import { useLocalizations } from "../../localization/makeLocalizable.js";
import { mqu } from "../../styledSystem/breakpoints.js";
import { useFormControl } from "../../utils/useFormControl.js";
import { useCardState, withCardStateProvider } from "../../elements/contexts/index.js";
import { Col, Flex, Text } from "../../customizables/index.js";
import { handleError } from "../../utils/errorHandler.js";
import { Form } from "../../elements/Form.js";
import { FormButtons } from "../../elements/FormButtons.js";
import { FormContainer } from "../../elements/FormContainer.js";
import { useActionContext } from "../../elements/Action/ActionRoot.js";
import { Action } from "../../elements/Action/index.js";
import { ThreeDotsMenu } from "../../elements/ThreeDotsMenu.js";
import { ProfileSection } from "../../elements/Section.js";
import { RemovePasskeyForm } from "./RemoveResourceForm.js";
import { getRelativeToNowDateKey } from "../../utils/getRelativeToNowDateKey.js";
import React, { Fragment, useState } from "react";
import { useClerk, useReverification, useUser } from "@clerk/shared/react";
import { jsx, jsxs } from "@emotion/react/jsx-runtime";

//#region src/components/UserProfile/PasskeySection.tsx
const RemovePasskeyScreen = (props) => {
	const { close } = useActionContext();
	return /* @__PURE__ */ jsx(RemovePasskeyForm, {
		onSuccess: close,
		onReset: close,
		...props
	});
};
const PasskeyScreen = (props) => {
	const { close } = useActionContext();
	return /* @__PURE__ */ jsx(UpdatePasskeyForm, {
		onSuccess: close,
		onReset: close,
		passkey: props.passkey
	});
};
const UpdatePasskeyForm = withCardStateProvider((props) => {
	const { onSuccess, onReset, passkey } = props;
	const card = useCardState();
	const passkeyNameField = useFormControl("passkeyName", passkey.name || "", {
		type: "text",
		label: localizationKeys("formFieldLabel__passkeyName"),
		isRequired: true
	});
	const canSubmit = passkeyNameField.value.length > 1 && passkey.name !== passkeyNameField.value;
	const addEmail = async (e) => {
		e.preventDefault();
		return passkey.update({ name: passkeyNameField.value }).then(onSuccess).catch((e$1) => handleError(e$1, [passkeyNameField], card.setError));
	};
	return /* @__PURE__ */ jsx(FormContainer, {
		headerTitle: localizationKeys("userProfile.passkeyScreen.title__rename"),
		headerSubtitle: localizationKeys("userProfile.passkeyScreen.subtitle__rename"),
		children: /* @__PURE__ */ jsxs(Form.Root, {
			onSubmit: addEmail,
			children: [/* @__PURE__ */ jsx(Form.ControlRow, {
				elementId: passkeyNameField.id,
				children: /* @__PURE__ */ jsx(Form.PlainInput, {
					...passkeyNameField.props,
					autoComplete: "off"
				})
			}), /* @__PURE__ */ jsx(FormButtons, {
				submitLabel: localizationKeys("userProfile.formButtonPrimary__save"),
				isDisabled: !canSubmit,
				onReset
			})]
		})
	});
});
const PasskeySection = () => {
	const { user } = useUser();
	const [actionValue, setActionValue] = useState(null);
	if (!user) return null;
	return /* @__PURE__ */ jsx(ProfileSection.Root, {
		title: localizationKeys("userProfile.start.passkeysSection.title"),
		centered: false,
		id: "passkeys",
		children: /* @__PURE__ */ jsx(Action.Root, {
			value: actionValue,
			onChange: setActionValue,
			children: /* @__PURE__ */ jsxs(ProfileSection.ItemList, {
				id: "passkeys",
				children: [user.passkeys.map((passkey) => {
					const passkeyId = passkey.id;
					return /* @__PURE__ */ jsxs(Fragment, { children: [
						/* @__PURE__ */ jsx(PasskeyItem, { ...passkey }, passkeyId),
						/* @__PURE__ */ jsx(Action.Open, {
							value: `remove-${passkeyId}`,
							children: /* @__PURE__ */ jsx(Action.Card, {
								variant: "destructive",
								children: /* @__PURE__ */ jsx(RemovePasskeyScreen, { passkey })
							})
						}),
						/* @__PURE__ */ jsx(Action.Open, {
							value: `rename-${passkeyId}`,
							children: /* @__PURE__ */ jsx(Action.Card, { children: /* @__PURE__ */ jsx(PasskeyScreen, { passkey }) })
						})
					] }, passkeyId);
				}), /* @__PURE__ */ jsx(AddPasskeyButton, { onClick: () => setActionValue(null) })]
			})
		})
	});
};
const PasskeyItem = (props) => {
	return /* @__PURE__ */ jsxs(ProfileSection.Item, {
		id: "passkeys",
		hoverable: true,
		sx: { alignItems: "flex-start" },
		children: [/* @__PURE__ */ jsx(PasskeyInfo, { ...props }), /* @__PURE__ */ jsx(ActiveDeviceMenu, { passkey: props })]
	});
};
const PasskeyInfo = (props) => {
	const { name, createdAt, lastUsedAt } = props;
	const { t } = useLocalizations();
	return /* @__PURE__ */ jsx(Flex, {
		sx: (t$1) => ({
			width: "100%",
			overflow: "hidden",
			gap: t$1.space.$4,
			[mqu.sm]: { gap: t$1.space.$2 }
		}),
		children: /* @__PURE__ */ jsxs(Col, {
			align: "start",
			gap: 1,
			children: [
				/* @__PURE__ */ jsx(Text, { children: name }),
				/* @__PURE__ */ jsxs(Text, {
					colorScheme: "secondary",
					children: ["Created: ", t(getRelativeToNowDateKey(createdAt))]
				}),
				lastUsedAt && /* @__PURE__ */ jsxs(Text, {
					colorScheme: "secondary",
					children: ["Last used: ", t(getRelativeToNowDateKey(lastUsedAt))]
				})
			]
		})
	});
};
const ActiveDeviceMenu = ({ passkey }) => {
	const { open } = useActionContext();
	const passkeyId = passkey.id;
	return /* @__PURE__ */ jsx(ThreeDotsMenu, { actions: [{
		label: localizationKeys("userProfile.start.passkeysSection.menuAction__rename"),
		onClick: () => open(`rename-${passkeyId}`)
	}, {
		label: localizationKeys("userProfile.start.passkeysSection.menuAction__destructive"),
		isDestructive: true,
		onClick: () => open(`remove-${passkeyId}`)
	}] });
};
const AddPasskeyButton = ({ onClick }) => {
	const card = useCardState();
	const { isSatellite } = useClerk();
	const { user } = useUser();
	const createPasskey = useReverification(() => user?.createPasskey());
	const handleCreatePasskey = async () => {
		onClick?.();
		if (!user) return;
		try {
			await createPasskey();
		} catch (e) {
			handleError(e, [], card.setError);
		}
	};
	if (isSatellite) return null;
	return /* @__PURE__ */ jsx(ProfileSection.ArrowButton, {
		id: "passkeys",
		localizationKey: localizationKeys("userProfile.start.passkeysSection.primaryButton"),
		onClick: handleCreatePasskey
	});
};

//#endregion
export { PasskeySection };
