// @clerk/ui/dist/no-rhc/elements/DevModeNotice.js
import { Box, Text } from "../customizables/index.js";
import { useDevMode } from "../hooks/useDevMode.js";
import { jsx } from "@emotion/react/jsx-runtime";

//#region src/elements/DevModeNotice.tsx
const DevModeOverlay = (props) => {
	const { gradient = 60 } = props;
	const { showDevModeNotice } = useDevMode();
	if (!showDevModeNotice) return null;
	return /* @__PURE__ */ jsx(Box, { sx: (t) => ({
		userSelect: "none",
		pointerEvents: "none",
		inset: 0,
		position: "absolute",
		background: `repeating-linear-gradient(-45deg,${t.colors.$warningAlpha100},${t.colors.$warningAlpha100} 6px,${t.colors.$warningAlpha150} 6px,${t.colors.$warningAlpha150} 12px)`,
		maskImage: `linear-gradient(transparent ${gradient}%, black)`
	}) });
};
const DevModeNotice = (props) => {
	const { sx } = props;
	const { showDevModeNotice } = useDevMode();
	if (!showDevModeNotice) return null;
	return /* @__PURE__ */ jsx(Text, {
		sx: [(t) => ({
			color: t.colors.$warning500,
			fontWeight: t.fontWeights.$semibold,
			padding: t.space.$1x5
		}), sx],
		children: "Development mode"
	});
};

//#endregion
export { DevModeNotice, DevModeOverlay };
